### PR TITLE
Implemented Power Split and Guard Split

### PR DESCRIPTION
**Pull request by XGSleepWalker**
* Implemented Power Split
* Implemented Guard Split

### CODE REVIEW

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -543,7 +543,12 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
   getStat(stat: Stat): integer {
-    return this.stats[stat];
+    if (!this.summonData)
+    {
+      return this.stats[stat];
+    }
+
+    return this.summonData.stats[stat];
```
Comment (bennybroseph) [inline src/field/pokemon.ts:551]: We use Egyptian style braces in this project, let's get these updated

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -1552,6 +1557,11 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
     return healAmount;
   }
 
+  changeSummonStat(stat: Stat, value: integer) : void
+  {
+    this.summonData.stats[stat] = value;
+  }
```
Comment (bennybroseph) [inline src/field/pokemon.ts:1563]: We use Egyptian style braces in this project, let's get these updated

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/move.ts [hunk 3]
```diff
@@ -1755,6 +1755,59 @@ export class HpSplitAttr extends MoveEffectAttr {
...
+      target.changeStat(Stat.SPDEF, specialDefenseValue);
+      infoUpdates.push(target.updateInfo());
+
+
+      return Promise.all(infoUpdates).then(() => resolve(true));
+    });
+  }
+
+}
+
```
[the remark this answers is not in the document]
Comment (XGSleepWalker) [reply src/data/move.ts:1921]: Documentation added

File: src/field/pokemon.ts (hunk 1 above)
Comment (XGSleepWalker) [reply src/field/pokemon.ts:552]: Braces fixed.

File: src/field/pokemon.ts (hunk 2 above)
Comment (XGSleepWalker) [reply src/field/pokemon.ts:1712]: Braces fixed.

**Review by bennybroseph — APPROVED** [review]

File: src/data/terrain.ts
Comment (bennybroseph) [inline src/data/terrain.ts:1]: Is this change needed?

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/move.ts [hunk 4]
```diff
@@ -1842,6 +1842,83 @@ export class HpSplitAttr extends MoveEffectAttr {
...
+   * @returns {boolean} True if power split is applied successfully.
+   */
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): Promise<boolean> {
+    return new Promise(resolve => {
+      if (!super.apply(user, target, move, args))
+        return resolve(false);
+
+      const infoUpdates = [];
+
+      const defenseValue = Math.floor((target.getStat(Stat.DEF) + user.getStat(Stat.DEF)) / 2);
```
Comment (bennybroseph) [inline src/data/move.ts:1903]: Is there a source that states these values are rounded down? If not we should be using Math.round here.

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/move.ts (hunk 4 above)
Comment (XGSleepWalker) [reply src/data/move.ts:1903]: From Smogon: https://www.smogon.com/dex/sv/moves/guard-split/
https://www.smogon.com/dex/sv/moves/power-split/

'The user and the target have their Attack/Defense and Special Attack/Defense stats set to be equal to the average of the user and the target's Attack and Special Attack stats, respectively, rounded down. Stat stage changes are unaffected.'

File: src/data/terrain.ts
Comment (XGSleepWalker) [reply src/data/terrain.ts:1]: Leftover from a bad push, removed.

File: src/field/pokemon.ts (hunk 2 above)
Comment (bennybroseph) [reply src/field/pokemon.ts:1712]: Didn't realize these were brand new functions. need these documented

File: src/field/pokemon.ts (hunk 1 above)
Comment (bennybroseph) [reply src/field/pokemon.ts:552]: Didn't realize these were functions. need these documented

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/field/pokemon.ts (hunk 1 above)
Comment (XGSleepWalker) [reply src/field/pokemon.ts:557]: Documentation added.

File: src/field/pokemon.ts (hunk 2 above)
Comment (XGSleepWalker) [reply src/field/pokemon.ts:1722]: Documentation added.

**Review by bennybroseph — APPROVED** [review]

**Pull request merged**